File: src/rendering/TextureRenderer.ts
import Logger from "js-logger";
import { autorun } from "mobx";
import {
    Mesh,
    MeshBasicMaterial,
    OrthographicCamera,
    PlaneGeometry,
    Texture,
    Vector2,
    Vector3,
} from "three";
import { Xvm } from "../data_formats/parsing/ninja/texture";
import { texture_viewer_store } from "../stores/TextureViewerStore";
import { Renderer } from "./Renderer";
import { xvm_texture_to_texture } from "./conversion/ninja_textures";

const logger = Logger.get("rendering/TextureRenderer");

let renderer: TextureRenderer | undefined;

export function get_texture_renderer(): TextureRenderer {
    if (!renderer) renderer = new TextureRenderer();
    return renderer;
}

export class TextureRenderer extends Renderer<OrthographicCamera> {
    private quad_meshes: Mesh[] = [];

    constructor() {
        super(new OrthographicCamera(-400, 400, 300, -300, 1, 10));

        this.controls.enableRotate = false;

        autorun(() => {
            this.scene.remove(...this.quad_meshes);

            const xvm = texture_viewer_store.current_xvm;

            if (xvm) {
                this.render_textures(xvm);
            }

            this.reset_camera(new Vector3(0, 0, 5), new Vector3());
            this.schedule_render();
        });
    }

    set_size(width: number, height: number): void {
        this.camera.left = -Math.floor(width / 2);
        this.camera.right = Math.ceil(width / 2);
        this.camera.top = Math.floor(height / 2);
        this.camera.bottom = -Math.ceil(height / 2);
        this.camera.updateProjectionMatrix();
        super.set_size(width, height);
    }

    private render_textures = (xvm: Xvm) => {
        let total_width = 10 * (xvm.textures.length - 1); // 10px spacing between textures.
        let total_height = 0;

        for (const tex of xvm.textures) {
            total_width += tex.width;
            total_height = Math.max(total_height, tex.height);
        }

        let x = -Math.floor(total_width / 2);
        const y = -Math.floor(total_height / 2);

        for (const tex of xvm.textures) {
            let tex_3js: Texture | undefined;

            try {
                tex_3js = xvm_texture_to_texture(tex);
            } catch (e) {
                logger.warn("Couldn't convert XVM texture.", e);
            }

            const quad_mesh = new Mesh(
                this.create_quad(
                    x,
                    y + Math.floor((total_height - tex.height) / 2),
                    tex.width,
                    tex.height
                ),
                tex_3js
                    ? new MeshBasicMaterial({
                          map: tex_3js,
                          transparent: true,
                      })
                    : new MeshBasicMaterial({
                          color: 0xff00ff,
                      })
            );

            this.quad_meshes.push(quad_mesh);
            this.scene.add(quad_mesh);

            x += 10 + tex.width;
        }
    };

    private create_quad(x: number, y: number, width: number, height: number): PlaneGeometry {
        const quad = new PlaneGeometry(width, height, 1, 1);
        quad.faceVertexUvs = [
            [
                [new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 0)],
                [new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0)],
            ],
        ];
        quad.translate(x + width / 2, y + height / 2, -5);
        return quad;
    }
}

File: web/src/main/kotlin/world/phantasmal/web/viewer/rendering/TextureRenderer.kt
package world.phantasmal.web.viewer.rendering

import mu.KotlinLogging
import org.khronos.webgl.Float32Array
import org.khronos.webgl.Uint16Array
import org.w3c.dom.HTMLCanvasElement
import world.phantasmal.psolib.fileFormats.ninja.XvrTexture
import world.phantasmal.web.core.rendering.*
import world.phantasmal.web.core.rendering.Renderer
import world.phantasmal.web.core.rendering.conversion.xvrTextureToThree
import world.phantasmal.web.externals.three.*
import world.phantasmal.web.viewer.stores.ViewerStore
import world.phantasmal.webui.obj
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.sqrt

private val logger = KotlinLogging.logger {}

class TextureRenderer(
    store: ViewerStore,
    createThreeRenderer: (HTMLCanvasElement) -> DisposableThreeRenderer,
) : Renderer() {
    private var meshes = listOf<Mesh>()

    override val context = addDisposable(RenderContext(
        createCanvas(),
        OrthographicCamera(
            left = -400.0,
            right = 400.0,
            top = 300.0,
            bottom = -300.0,
            near = 1.0,
            far = 10.0,
        )
    ))

    override val threeRenderer = addDisposable(createThreeRenderer(context.canvas)).renderer

    override val inputManager = addDisposable(OrbitalCameraInputManager(
        context.canvas,
        context.camera,
        Vector3(0.0, 0.0, 5.0),
        screenSpacePanning = true,
        enableRotate = false,
    ))

    init {
        observeNow(store.currentTextures) {
            texturesChanged(it.filterNotNull())
        }
    }

    private fun texturesChanged(textures: List<XvrTexture>) {
        meshes.forEach { mesh ->
            context.scene.remove(mesh)
            disposeObject3DResources(mesh)
        }

        inputManager.resetCamera()

        // Lay textures out in a square grid of "cells".
        var cellWidth = -1
        var cellHeight = -1

        textures.forEach {
            cellWidth = max(cellWidth, SPACING + it.width)
            cellHeight = max(cellHeight, SPACING + it.height)
        }

        val cellsPerRow = ceil(sqrt(textures.size.toDouble())).toInt()
        val cellsPerCol = ceil(textures.size.toDouble() / cellsPerRow).toInt()

        // Start at the center of the first cell because the texture quads are placed at the center
        // of the given coordinates.
        val startX = -(cellsPerRow * cellWidth) / 2 + cellWidth / 2
        var x = startX
        var y = (cellsPerCol * cellHeight) / 2 - cellHeight / 2
        var cell = 0

        meshes = textures.map { xvr ->
            val texture =
                try {
                    xvrTextureToThree(xvr, filter = NearestFilter)
                } catch (e: Exception) {
                    logger.error(e) { "Couldn't convert XVR texture." }
                    null
                }

            val quad = Mesh(
                createQuad(x, y, xvr.width, xvr.height),
                MeshBasicMaterial(obj {
                    if (texture == null) {
                        color = Color(0xFF00FF)
                    } else {
                        map = texture
                        transparent = true
                    }
                })
            )
            context.scene.add(quad)

            x += cellWidth

            if (++cell % cellsPerRow == 0) {
                x = startX
                y -= cellHeight
            }

            quad
        }
    }

    private fun createQuad(x: Int, y: Int, width: Int, height: Int): BufferGeometry {
        val halfWidth = width / 2f
        val halfHeight = height / 2f

        val geom = BufferGeometry()

        geom.setAttribute(
            "position",
            Float32BufferAttribute(
                Float32Array(arrayOf(
                    -halfWidth, -halfHeight, 0f,
                    -halfWidth, halfHeight, 0f,
                    halfWidth, halfHeight, 0f,
                    halfWidth, -halfHeight, 0f,
                )),
                3,
            ),
        )
        geom.setAttribute(
            "normal",
            Float32BufferAttribute(
                Float32Array(arrayOf(
                    0f, 0f, 1f,
                    0f, 0f, 1f,
                    0f, 0f, 1f,
                    0f, 0f, 1f,
                )),
                3,
            ),
        )
        geom.setAttribute(
            "uv",
            Float32BufferAttribute(
                Float32Array(arrayOf(
                    0f, 1f,
                    0f, 0f,
                    1f, 0f,
                    1f, 1f,
                )),
                2,
            ),
        )
        geom.setIndex(Uint16BufferAttribute(
            Uint16Array(arrayOf(
                0, 2, 1,
                2, 0, 3,
            )),
            1,
        ))

        geom.translate(x.toDouble(), y.toDouble(), -5.0)

        return geom
    }

    companion object {
        private const val SPACING = 10
    }
}
